File: src/2022-intern/763.partition-labels.js
/*
 * @lc app=leetcode id=763 lang=javascript
 *
 * [763] Partition Labels
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number[]}
 */
const partitionLabels = function (s) {
  /**
   * 某一组的最后一个字母的下标
   * @type {number[]} */
  const splitIndexes = [-1];

  /**
   * 某个字母的最后一个下标
   * @type {Map<string, number>}
   */
  const letterLastIndexes = new Map();

  for (let i = 0; i < s.length; i++) {
    letterLastIndexes.set(s[i], i);
  }

  for (let i = 0; i < s.length; i++) {
    // 新的一组
    if (i > splitIndexes[splitIndexes.length - 1]) {
      splitIndexes.push(i);
    }

    const letterLastIndex = letterLastIndexes.get(s[i]);
    // 扩充当前分组
    if (letterLastIndex > splitIndexes[splitIndexes.length - 1]) {
      splitIndexes[splitIndexes.length - 1] = letterLastIndex;
    }
  }

  const result = [];
  for (let i = 0; i < splitIndexes.length - 1; i++) {
    result.push(splitIndexes[i + 1] - splitIndexes[i]);
  }

  return result;
};
// @lc code=end

File: src/2025/763.partition-labels.cpp
/*
 * @lc app=leetcode id=763 lang=cpp
 *
 * [763] Partition Labels
 */

#include <algorithm>
#include <array>
#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<int> partitionLabels(const std::string& s) {
    std::array<int, 26> letter_last_indexes{};
    std::fill(letter_last_indexes.begin(), letter_last_indexes.end(), -1);
    for (int i = 0; i < s.size(); i++) {
      letter_last_indexes[s[i]] = i;
    }

    int current_partition_begin = -1;
    int current_partition_end = -1;
    std::vector<int> partition_sizes;
    for (int i = 0; i < s.size(); i++) {
      if (current_partition_begin == -1) {
        current_partition_begin = i;
        current_partition_end = i;
      }
      if (i > current_partition_end) {
        partition_sizes.push_back(current_partition_end -
                                  current_partition_begin + 1);
        current_partition_begin = i;
      }
      current_partition_end =
          std::max(current_partition_end, letter_last_indexes[s[i]]);
    }
    partition_sizes.push_back(current_partition_end - current_partition_begin +
                              1);
    return partition_sizes;
  }
};
// @lc code=end
